mscore/palette.cpp

```diff
@@ -730,12 +765,306 @@ void Palette::leaveEvent(QEvent*)
       {
       if (currentIdx != -1) {
             QRect r = idxRect(currentIdx);
-            currentIdx = -1;
+            PaletteBox* pb = mscore->getPaletteBox();
+            if (!filterActive || !pb->getKeyboardNavigation())
+                  currentIdx = -1;
             update(r);
             }
       }
 
 //---------------------------------------------------------
+//   nextPaletteElement
+//---------------------------------------------------------
+void Palette::nextPaletteElement()
+      {
+      if (!filterActive)
+            return;
+      int i = currentIdx;
+      if (i == -1)
+            return;
+      i++;
+      if (i < size() && cellAt(i)) {
+            currentIdx = i;
+            }
+      update();
+      }
+
+//---------------------------------------------------------
+//   prevPaletteElement
+//---------------------------------------------------------
+
+void Palette::prevPaletteElement()
+      {
+      if (!filterActive)
+            return;
+      int i = currentIdx;
+      if (i == -1)
+            return;
+      i--;
+      if (i >= 0 && cellAt(i)) {
+            currentIdx = i;
+            }
+      update();
+      }
+
+//---------------------------------------------------------
+//   applyPaletteElement
+//---------------------------------------------------------
+
+void Palette::applyPaletteElement()
```

Comment:
this is largely the same as mosueDoubleClickedEvent.  Is there a way to refactor the code to have a single function that does most of the work for both?